### PR TITLE
feat: flexible init scaffolding

### PR DESCRIPTION
**Pull request by anshumanv**
**What kind of change does this PR introduce?**
feat

**Did you add tests for your changes?**
not yet

**If relevant, did you update the documentation?**
NA

**Summary**
- Gives option for users to select plugins they want instead of applying them by default

**Does this PR introduce a breaking change?**
No

**Other information**
/cc @alexander-akait

### CODE REVIEW

File: packages/generators/src/init-generator.ts [hunk 1]
```diff
@@ -201,7 +201,24 @@ export default class InitGenerator extends CustomGenerator {
...
             });
         }
-        if (this.usingDefaults) {
+
+        // webpack Dev Server
+        const { useDevServer } = await Confirm(self, 'useDevServer', 'Do you want to use webpack-dev-server', true, false);
+        if (useDevServer) {
+            this.dependencies.push('webpack-dev-server');
+            this.configuration.config.webpackOptions.devServer = {
+                open: true,
```
Comment (alexander-akait) [inline packages/generators/src/init-generator.ts:210]: Let's add host: `localhost` too, just for example

Comment (anshumanv) [reply packages/generators/src/init-generator.ts:210]: yep lets do it, I think it's fine to make `localhost` as default

**Review by rishabh3112 — COMMENTED** [review]
Looks good.
Lets update https://github.com/webpack/webpack-cli/blob/master/INIT.md as well.
Left few suggestions as well.

File: packages/generators/src/init-generator.ts [hunk 2]
```diff
@@ -201,7 +201,24 @@ export default class InitGenerator extends CustomGenerator {
...
+            };
+        }
+
+        const { useHTMLPlugin } = await Confirm(
+            self,
+            'useHTMLPlugin',
+            'Do you want to use html-webpack-plugin to simplify creation of HTML files for your bundle?',
```
[the remark this answers is not in the document]
Comment (anshumanv) [reply packages/generators/src/init-generator.ts:217]: yeah this makes sense, but it's ambiguous to the user what we'll install behind the scenes, can we think of something better?

[the remark this answers is not in the document]
Comment (rishabh3112) [reply packages/generators/src/init-generator.ts:217]: We already points users to https://github.com/webpack/webpack-cli/blob/master/INIT.md for getting more info. Should be fine I think.

File: packages/generators/src/init-generator.ts [hunk 3]
```diff
@@ -214,14 +231,17 @@ export default class InitGenerator extends CustomGenerator {
...
-            this.dependencies.push('webpack-dev-server');
-            this.configuration.config.webpackOptions.devServer = {
-                open: true,
-            };
-
-            // PWA + offline support
+        const { useWorkboxPlugin } = await Confirm(
+            self,
+            'useDevServer',
+            'Do you want to use workbox-webpack-plugin for PWA support?',
```
Comment (rishabh3112) [inline packages/generators/src/init-generator.ts:239]:
```suggestion
            'Do you want to add PWA support?',
```

**Review by rishabh3112 — COMMENTED** [review]
Another question. Dev Server question seems fine to me.

File: packages/generators/src/init-generator.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (anshumanv) [reply packages/generators/src/init-generator.ts:217]: sounds good, updating

**Comment by anshumanv** [comment]
Done changes @alexander-akait @rishabh3112

**Review by alexander-akait — DISMISSED** [review]

**Review by rishabh3112 — DISMISSED** [review]

File: INIT.md [hunk 4]
```diff
@@ -135,3 +135,17 @@ If you use any sort of style in your project, such as [`.less`](http://lesscss.o
...
 
 If you indicate based on previous questions that you are using production, this will be enabled. The default value for your generated CSS file is `style.[contentHash].css`, which will collect all your `.less`, `.scss` or `.css` into one file. This will make your build faster in production.
+
+7. `Do you want to use webpack-dev-server?`
+
+> _Property/key resolved: [module.rules](https://webpack.js.org/configuration/dev-server/)_
+
+Adds a development server to serve webpack bundles and hence make development faster.
+
+8. `Do you want to use simplify creation of HTML files for your bundle?`
```
Comment (jamesgeorge007) [inline INIT.md:145]:
```suggestion
8. `Do you want to simplify the creation of HTML files for your bundle?`
```

File: packages/generators/src/init-generator.ts [hunk 5]
```diff
@@ -201,7 +201,25 @@ export default class InitGenerator extends CustomGenerator {
...
+            this.configuration.config.webpackOptions.devServer = {
+                open: true,
+                host: 'localhost',
+            };
+        }
+
+        const { useHTMLPlugin } = await Confirm(
+            self,
+            'useHTMLPlugin',
+            'Do you want to use simplify creation of HTML files for your bundle?',
```
Comment (jamesgeorge007) [inline packages/generators/src/init-generator.ts:218]:
```suggestion
            'Do you want to simplify the creation of HTML files for your bundle?',
```

File: packages/generators/src/init-generator.ts [hunk 6]
```diff
@@ -201,7 +201,25 @@ export default class InitGenerator extends CustomGenerator {
                 use: ExtractUseProps,
             });
         }
-        if (this.usingDefaults) {
+
+        // webpack Dev Server
+        const { useDevServer } = await Confirm(self, 'useDevServer', 'Do you want to use webpack-dev-server?', true, false);
```
Comment (jamesgeorge007) [inline packages/generators/src/init-generator.ts:206]:
```suggestion
        const { useDevServer } = await Confirm(self, 'useDevServer', 'Do you want to use webpack-dev-server?', true, this.usingDefaults);
```

File: packages/generators/src/init-generator.ts [hunk 7]
```diff
@@ -201,7 +201,25 @@ export default class InitGenerator extends CustomGenerator {
...
+                host: 'localhost',
+            };
+        }
+
+        const { useHTMLPlugin } = await Confirm(
+            self,
+            'useHTMLPlugin',
+            'Do you want to simplify the creation of HTML files for your bundle?',
+            false,
+            false,
```
Comment (jamesgeorge007) [inline packages/generators/src/init-generator.ts:220]:
```suggestion
            this.usingDefaults,
```

File: packages/generators/src/init-generator.ts [hunk 8]
```diff
@@ -214,14 +232,11 @@ export default class InitGenerator extends CustomGenerator {
...
+        }
 
-            // webpack Dev Server
-            this.dependencies.push('webpack-dev-server');
-            this.configuration.config.webpackOptions.devServer = {
-                open: true,
-            };
-
-            // PWA + offline support
+        const { useWorkboxPlugin } = await Confirm(self, 'useDevServer', 'Do you want to add PWA support?', true, false);
```
Comment (jamesgeorge007) [inline packages/generators/src/init-generator.ts:237]:
```suggestion
        const { useWorkboxPlugin } = await Confirm(self, 'useDevServer', 'Do you want to add PWA support?', true, this.usingDefaults);
```

**Review by jamesgeorge007 — CHANGES_REQUESTED** [review]
`webpack init --auto` is supposed to go ahead with the default values; however, these prompts show up currently.

https://github.com/webpack/webpack-cli/blob/e44e855c7e302932a828fcedf7abfe205b47c716/packages/generators/src/init-generator.ts#L186

**Comment by alexander-akait** [comment]
Can we improve coverage?

**Comment by anshumanv** [comment]
Updated tests to already cover new cases

**Review by alexander-akait — APPROVED** [review]

**Review by snitin315 — APPROVED** [review]

**Comment by alexander-akait** [comment]
@webpack/cli-team Please review, we need merge this and continue fixing

**Review by jamesgeorge007 — APPROVED** [review]

**Pull request merged**